packages/runtime/src/webcontainer/preview-info.ts

```diff
@@ -15,27 +15,41 @@ export class PreviewInfo {
     return undefined;
   }
 
-  constructor(preview: Exclude<PreviewSchema, boolean>[0], ready?: boolean) {
-    if (typeof preview === 'number') {
-      this.port = preview;
+  get port() {
+    return this.portInfo.port;
+  }
+
+  get baseUrl() {
+    return this.portInfo.origin;
+  }
+
+  get ready() {
+    return this.portInfo.ready;
+  }
+
+  constructor(preview: PortInfo | Exclude<PreviewSchema, boolean>[0], ready?: boolean) {
+    if (preview instanceof PortInfo) {
+      this.portInfo = preview;
+    } else if (typeof preview === 'number') {
+      this.portInfo = new PortInfo(preview);
     } else if (typeof preview === 'string') {
       const [port, ...rest] = preview.split('/');
-      this.port = parseInt(port);
+      this.portInfo = new PortInfo(parseInt(port));
       this.pathname = rest.join('/');
     } else if (Array.isArray(preview)) {
-      this.port = preview[0];
+      this.portInfo = new PortInfo(preview[0]);
       this.title = preview[1];
       this.pathname = preview[2];
     } else {
-      this.port = preview.port;
+      this.portInfo = new PortInfo(preview.port);
       this.title = preview.title;
       this.pathname = preview.pathname;
```

Comment:
I think we probably want to move out the parsing logic outside of the constructor. The constructor could accept everything as arguments and we have a static method that does the parsing and returns `{ port: number; pathname?: string; title?: string }` as an object.

Reply:
Added `PreviewInfo.parse` that's used to parse the schema into preview's properties.